### PR TITLE
feat (api): add release builtin action

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: engine/api/repositoriesmanager/repogithub/client.go [hunk 1]
```diff
@@ -22,6 +24,76 @@ type GithubClient struct {
...
+		Body:    releaseNote,
+	}
+	b, err := json.Marshal(req)
+	if err != nil {
+		return nil, sdk.WrapError(err, "github.Release > Cannot marshal body %+v", req)
+	}
+
+	res, err := g.post(url, "application/json", bytes.NewBuffer(b), false)
+	if err != nil {
+		return nil, err
```
Comment (fsamin) [inline engine/api/repositoriesmanager/repogithub/client.go]: wrap

File: engine/api/repositoriesmanager/repostash/client.go [hunk 2]
```diff
@@ -34,6 +35,16 @@ type StashClient struct {
 	disableSetStatus bool
 }
 
+// Release not implemented on bitbucket
+func (s *StashClient) Release(repo string, tagName string, title string, releaseNote string) (*sdk.VCSRelease, error) {
+	return nil, fmt.Errorf("Stash do not provide release system")
```
Comment (fsamin) [inline engine/api/repositoriesmanager/repostash/client.go]: en fait, il faudrait que la release soit faite sur nexus...

File: engine/api/workflow_application.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package main
+
+import (
+	"net/http"
+	"sort"
+
+	"github.com/go-gorp/gorp"
+	"github.com/gorilla/mux"
+
+	"bufio"
```
Comment (fsamin) [inline engine/api/workflow_application.go]: import order

File: engine/api/workflow_application.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	if errNRI != nil {
+		return errNRI
+	}
+
+	var req sdk.WorkflowNodeRunRelease
+	if errU := UnmarshalBody(r, &req); errU != nil {
+		return errU
+	}
+
+	router.Handle("/project/{permProjectKey}/workflows/{workflowName}/runs/{number}/nodes/{id}/release", POST(releaseApplicationWorkflowHandler))
```
Comment (fsamin) [inline engine/api/workflow_application.go]: ???

File: engine/api/workflow_application.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+			if ok {
+				artifactToUpload = append(artifactToUpload, a)
+				break
+			}
+		}
+	}
+
+	for _, a := range artifactToUpload {
+		b := &bytes.Buffer{}
+		writer := bufio.NewWriter(b)
```
Comment (fsamin) [inline engine/api/workflow_application.go]: bytes.Buffer is already a Writer

File: engine/api/action/builtin.go [hunk 6]
```diff
@@ -151,6 +151,37 @@ Tag the current branch and push it.`
...
+	gitrelease.Parameter(sdk.Parameter{
+		Name:        "title",
+		Value:       "",
+		Description: "Set a title for the release",
+		Type:        sdk.StringParameter,
+	})
+	gitrelease.Parameter(sdk.Parameter{
+		Name:        "releaseNote",
+		Description: "Set a release note for the release",
+		Type:        sdk.StringParameter,
```
Comment (yesnault) [inline engine/api/action/builtin.go]: TextParameter

File: engine/api/action/builtin.go [hunk 7]
```diff
@@ -151,6 +151,37 @@ Tag the current branch and push it.`
...
+	})
+	gitrelease.Parameter(sdk.Parameter{
+		Name:        "artifacts",
+		Description: "Set a list of artifacts, separate by , .",
```
Comment (yesnault) [inline engine/api/action/builtin.go]: + doc sur les regex

File: engine/api/action/builtin.go [hunk 8]
```diff
@@ -151,6 +151,37 @@ Tag the current branch and push it.`
...
+		Type:        sdk.StringParameter,
```
Comment (yesnault) [inline engine/api/action/builtin.go]: TextParameter

File: engine/api/repositoriesmanager/repogithub/client.go [hunk 9]
```diff
@@ -22,6 +24,76 @@ type GithubClient struct {
...
+
+// ReleaseResponse Response return by Github after release creation
+type ReleaseResponse struct {
+	ID        int64  `json:"id"`
+	UploadURL string `json:"upload_url"`
+}
+
+// Release Create a release Github
+func (g *GithubClient) Release(fullname string, tagName string, title string, releaseNote string) (*sdk.VCSRelease, error) {
+	var url = "/repos/" + fullname + "/releases"
```
Comment (yesnault) [inline engine/api/repositoriesmanager/repogithub/client.go]: fullname needs escape ?

File: sdk/cdsclient/client_workflow.go [hunk 10]
```diff
@@ -74,3 +74,15 @@ func (c *client) WorkflowNodeRunArtifactDownload(projectKey string, name string,
...
 }
+
+func (c *client) WorkflowNodeRunRelease(projectKey string, workflowName string, runNumber int64, nodeRunID int64, release sdk.WorkflowNodeRunRelease) error {
+	url := fmt.Sprintf("/project/%s/workflows/%s/runs/%d/nodes/%d/release", projectKey, workflowName, runNumber, nodeRunID)
+	code, err := c.PostJSON(url, release, nil)
+	if err != nil {
+		return err
+	}
+	if code >= 300 {
+		return fmt.Errorf("Cannot create workflow node run release")
```
Comment (yesnault) [inline sdk/cdsclient/client_workflow.go]: display http code ? (and body err ?)

File: engine/api/workflow/process_parameters.go [hunk 11]
```diff
@@ -80,20 +76,22 @@ func GetNodeBuildParameters(proj *sdk.Project, w *sdk.Workflow, n *sdk.WorkflowN
...
 	}
 
-	tmp["cds.project"] = w.ProjectKey
-	tmp["cds.workflow"] = w.Name
-	tmp["cds.pipeline"] = n.Pipeline.Name
-	tmp["cds.version"] = fmt.Sprintf("%d.%d", 1, 0)
-	tmp["cds.run"] = fmt.Sprintf("%d.%d", 1, 0)
-	tmp["cds.run.number"] = fmt.Sprintf("%d", 1)
-	tmp["cds.run.subnumber"] = fmt.Sprintf("%d", 0)
+	// TODO Update suggest.go  with new variable
```
Comment (bnjjj) [inline engine/api/workflow/process_parameters.go]: Don't forget

File: engine/api/repositoriesmanager/repogithub/client.go (hunk 9 above)
Comment (sguiheux) [reply engine/api/repositoriesmanager/repogithub/client.go]: no

File: engine/api/workflow/process_parameters.go (hunk 11 above)
Comment (sguiheux) [reply engine/api/workflow/process_parameters.go]: it's not for me ;)

**Pull request merged**